### PR TITLE
rgw: extend automatic rgw pool creation capability

### PR DESCRIPTION
**Pull request by alimaredia**
Signed-off-by: Ali Maredia <amaredia@redhat.com>

### CODE REVIEW

File: group_vars/rgws.yml.sample [hunk 1]
```diff
@@ -33,20 +33,46 @@ dummy:
...
+# pools declared in this dictionary.
+# The default value of {{ rgw_zone }} is 'default'.
+# The pool_type must be set as either 'replicated' or 'ec' for
+# each pool. 
+# If a pool's pool_type is 'ec', k and m values must be set via
+# the ec_k, and ec_m variables.
 #rgw_create_pools:
-#  defaults.rgw.buckets.data:
-#    pg_num: 16
+#  {{ rgw_zone }}.rgw.buckets.data:
```
Comment (dsavineau) [inline group_vars/rgws.yml.sample]: this needs to be quoted (and all the other keys using a variable)

File: roles/ceph-rgw/defaults/main.yml [hunk 2]
```diff
@@ -25,20 +25,46 @@ copy_admin_key: false
...
+# pools declared in this dictionary.
+# The default value of {{ rgw_zone }} is 'default'.
+# The pool_type must be set as either 'replicated' or 'ec' for
+# each pool. 
+# If a pool's pool_type is 'ec', k and m values must be set via
+# the ec_k, and ec_m variables.
 #rgw_create_pools:
-#  defaults.rgw.buckets.data:
-#    pg_num: 16
+#  {{ rgw_zone }}.rgw.buckets.data:
```
Comment (dsavineau) [inline roles/ceph-rgw/defaults/main.yml]: same here

File: roles/ceph-rgw/tasks/rgw_create_pools.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
+---
+  - name: remove ec profile
+    command: "{{ container_exec_cmd }} ceph --connect-timeout 5 --cluster {{ cluster }} osd erasure-code-profile rm {{ item.ec_profile }}"
```
Comment (dsavineau) [inline roles/ceph-rgw/tasks/rgw_create_pools.yml]: because we're using a dict lookup the values need to use `item.value.<value>`
So it should be `item.value.ec_profile`

File: roles/ceph-rgw/tasks/rgw_create_pools.yml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    delegate_to: "{{ groups[mon_group_name][0] }}"
+    changed_when: false
+    run_once: true
+    register: result
+    until: result is succeeded
+    when: item.pool_type == 'ec'
+    ignore_errors: yes
+
+  - name: set ec profile
+    command: "{{ container_exec_cmd }} ceph --connect-timeout 5 --cluster {{ cluster }} osd erasure-code-profile set {{ item.ec_profile }} k={{ item.ec_k }} m={{ item.ec_m }}"
```
Comment (dsavineau) [inline roles/ceph-rgw/tasks/rgw_create_pools.yml]: same here

**Comment by guits** [comment]
@alimaredia when you have a moment :slightly_smiling_face:

File: roles/ceph-rgw/tasks/rgw_create_pools.yml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  - name: set ec profile
+    command: "{{ container_exec_cmd }} ceph --connect-timeout 5 --cluster {{ cluster }} osd erasure-code-profile set {{ item.value.ec_profile }} k={{ item.value.ec_k }} m={{ item.value.ec_m }}"
+    with_dict: "{{ rgw_create_pools }}"
+    delegate_to: "{{ groups[mon_group_name][0] }}"
+    changed_when: false
+    run_once: true
+    register: result
+    when: 
+      - item.value.pool_type is defined  
+      - item.value.pool_type == 'ec'
```
Comment (alimaredia) [inline roles/ceph-rgw/tasks/rgw_create_pools.yml]: @guits should I check to make sure `ec_k` and `ec_m` are defined before this job is run?

**Comment by alimaredia** [comment]
@guits @dsavineau: This PR should be ready for final review. I made some changes in the `tests` directory, let me know if there's anything else that needs to be done there.

File: roles/ceph-rgw/tasks/rgw_create_pools.yml [hunk 6]
```diff
@@ -0,0 +1,74 @@
+---
+  - name: remove ec profile
+    command: "{{ container_exec_cmd }} ceph --connect-timeout 5 --cluster {{ cluster }} osd erasure-code-profile rm {{ item.value.ec_profile }}"
+    with_dict: "{{ rgw_create_pools }}"
+    delegate_to: "{{ groups[mon_group_name][0] }}"
+    changed_when: false
+    run_once: true
+    register: result
```
Comment (guits) [inline roles/ceph-rgw/tasks/rgw_create_pools.yml]: all register here seem to be unused

File: roles/ceph-rgw/tasks/rgw_create_pools.yml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    when: 
+      - item.value.pool_type is defined  
+      - item.value.pool_type == 'ec'
+    ignore_errors: yes
```
Comment (guits) [inline roles/ceph-rgw/tasks/rgw_create_pools.yml]: why?

File: roles/ceph-rgw/tasks/rgw_create_pools.yml (hunk 5 above)
Comment (guits) [reply roles/ceph-rgw/tasks/rgw_create_pools.yml]: > @guits should I check to make sure `ec_k` and `ec_m` are defined before this job is run?

That wouldn't hurt :slightly_smiling_face: 

> Also is there a way to fail the job if those variables are not defined but `pool_type` is set to `ec`?

like an early check in ceph-validate?

File: roles/ceph-rgw/tasks/rgw_create_pools.yml (hunk 7 above)
Comment (alimaredia) [reply roles/ceph-rgw/tasks/rgw_create_pools.yml]: The reason we had this was if the playbook gets run again. When the playbook gets run again, it will try to rm the ec_profile and we didn't want it to fail if the ec_profile was already created and being used.

File: roles/ceph-rgw/tasks/rgw_create_pools.yml (hunk 5 above)
Comment (alimaredia) [reply roles/ceph-rgw/tasks/rgw_create_pools.yml]: Done. Take a look at the changes I just made to the ceph-validate task.

File: roles/ceph-rgw/tasks/main.yml [hunk 8]
```diff
@@ -6,6 +6,10 @@
   include_tasks: pre_requisite.yml
   when: not containerized_deployment | bool
 
+- name: rgw pool creation tasks
```
Comment (dsavineau) [inline roles/ceph-rgw/tasks/main.yml]: Why did you move the pools creation here ?

Shouldn't this be done after the multisite tasks ?

File: roles/ceph-rgw/tasks/rgw_create_pools.yml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    with_dict: "{{ rgw_create_pools }}"
+    delegate_to: "{{ groups[mon_group_name][0] }}"
+    changed_when: false
+    run_once: true
+    when: 
+      - item.value.pool_type is defined  
+      - item.value.pool_type == 'ec'
+
+  - name: set crush rule
+    command: "{{ container_exec_cmd }} ceph --connect-timeout 5 --cluster {{ cluster }} osd crush rule create-erasure {{ item.key }} {{ item.value.ec_profile }}"
```
Comment (guits) [inline roles/ceph-rgw/tasks/rgw_create_pools.yml]: what about this command?
If the playbook is run again, will it return 0 even though it's already created? Otherwise, we might need a `failed_when: false` here as well

File: roles/ceph-validate/tasks/check_rgw_pools.yml [hunk 10]
```diff
@@ -0,0 +1,44 @@
+---
+- name: fail if pool_type is not set
```
Comment (dsavineau) [inline roles/ceph-validate/tasks/check_rgw_pools.yml]: I think this is too restrictive and some tasks are breaking the backward compatibility.

As I mentioned in a previous comment, users that wouldn't have defined pool_type in the rgw_create_pools dict will have the playbook failing.

Some checks make sense, like checking ec_m/ec_k when the pool type is ec but others like size and pg_num have default value so you don't actually need to always define these keys in the dict (ie: `{{ item.value.pg_num | default(osd_pool_default_pg_num) }}`)

File: roles/ceph-rgw/tasks/main.yml (hunk 8 above)
Comment (alimaredia) [reply roles/ceph-rgw/tasks/main.yml]: @cbodley and I realized that pool creation should be happening before the rgw is starting up or any multisite tasks.

File: roles/ceph-rgw/tasks/rgw_create_pools.yml (hunk 9 above)
Comment (guits) [reply roles/ceph-rgw/tasks/rgw_create_pools.yml]: @alimaredia ?

**Comment by dsavineau** [comment]
Closing in flavor of #4688
